pandas/tests/arithmetic/test_datetime64.py

```diff
@@ -1435,27 +1435,38 @@ def test_dt64arr_add_sub_offset_ndarray(self, tz_naive_fixture,
         expected = tm.box_expected(expected, box_with_array)
         tm.assert_equal(res, expected)
 
+    @pytest.mark.parametrize("box", [pd.Index, pd.Series, pd.DataFrame])
```

Comment:
Can you use box_with_array fixture?

Reply:
In the case `to_array` of `box_with_array`, the DatetimeArray after addition of an offset of _month and day_ doesn't have any freq. But adding an offset of _month only_ results in freq. May I assert the equality of array values only by `assert_numpy_array_equal`?
```
In [18]: arr = pd.date_range(start='01 Jan 2014', end='01 Jan 2017', freq='AS').
    ...: array                                                                  

In [19]: arr                                                                    
Out[19]: 
<DatetimeArray>
['2014-01-01 00:00:00', '2015-01-01 00:00:00', '2016-01-01 00:00:00',
 '2017-01-01 00:00:00']
Length: 4, dtype: datetime64[ns]

In [20]: offset1 = pd.DateOffset(months=3, days=10)                             

In [21]: offset1                                                                
Out[21]: <DateOffset: days=10, months=3>

In [22]: (arr + offset1).freq is None                                           
Out[22]: True

In [23]: offset2 = pd.DateOffset(months=3)                                      

In [24]: (arr + offset2).freq                                                   
Out[24]: <YearBegin: month=4>
```

Reply:
> In the case to_array of box_with_array, the DatetimeArray after addition of an offset of month and day doesn't have any freq.

If DatetimeIndex + DateOffset behaves differently from DatetimeArray + DateOffset, that's a bug and should be fixed.

Reply:
The difference starts at L391:
https://github.com/pandas-dev/pandas/blob/d3a19121790a517c03755fc01f74ab149d9c77f5/pandas/core/arrays/datetimes.py#L381-L384
There is a frequency `AS-JAN` defined to denote the _beginning_ of Jan, but no frequency defined for a _specific_ date.
ref.: [https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#anchored-offsets](url)
```
In [2]: arr1 = pd.DatetimeIndex(['01 Jan 2014', '01 Jan 2015', '01 Jan 2016', '0
   ...: 1 Jan 2017']).array                                                     

In [3]: arr2 = pd.DatetimeIndex(['11 Apr 2014', '11 Apr 2015', '11 Apr 2016', '1
   ...: 1 Apr 2017']).array                                                     

In [4]: arr1.inferred_freq                                                      
Out[4]: 'AS-JAN'

In [5]: arr2.inferred_freq                                                      

In [6]: arr2.inferred_freq is None                                              
Out[6]: True
```
is it overhead to add more frequency definition?

Reply:
I think there was a miscommunication and I got confused.  Hopefully I've got it right now, let's double-check.  Thanks for bearing with me here.

The following should be true for any DatetimeIndex `dti` and any DateOffset `offset`:

```
(dti + offset).freq == (dti.array + offset).freq
```

Are you saying there is a counter-example?  If not, I don't see why you can't use `box_with_array`.

Reply:
`(dti + offset).freq == (dti.array + offset).freq`
This equality holds.

I add the freq check of the Index myself, because `assert_index_equal` doesn't check the freq. The freq of the Index match the expected frequency as the array does.

Reply:
Got it, thanks.

I think there is an Issue for having `assert_index_equal` check `freq` for appropriate subclasses, if that's something you're interested in fixing.